common/makefiles/docs/generic-make-go.md

```diff
@@ -0,0 +1,116 @@
+| build-image                       | Build a Docker image used in the CI environment.    ??                 |
+| push-image                        | Pushes the image to image registry, used in CI environment.           |
+
+### Add Generic Makefile
+To add generic Makefile, add to your Makefile following things and fill the variables.
+```makefile
+APP_NAME = {APPLICATION NAME}
+APP_PATH = {APPLICATION PATH IN REPOSITORY}
+BUILDPACK = {BUILDPACK IMAGE}
+SCRIPTS_DIR = {GENERIC MAKEFILE PATH} 
+include $(SCRIPTS_DIR)/generic-make-go.mk
+```
+Find the list of available images [here](https://github.com/kyma-project/test-infra/blob/master/templates/config.yaml).
+
+### Detailed workflow 
+By example:
+When CI run`make release` the following steps are executed:
+- rule `release` depends on rules `resolve dep-status verify build-image push-image`
+- rule`resolve` does not appear in the Makefile, but it's generated. 
+Notice this line of code:
+```makefile
+MOUNT_TARGETS = build resolve ensure dep-status check-imports imports check-fmt fmt errcheck vet generate pull-licenses gqlgen
+$(foreach t,$(MOUNT_TARGETS),$(eval $(call buildpack-mount,$(t))))
+```
+for resolve, and others targets defined in `MOUNT_TARGETS`, function `buildpack-mount` is callled, which dynamically defines new rule:
+```makefile
+resolve:
+    @echo make resolve
+    @docker run $(DOCKER_INTERACTIVE) \
+        -v $(COMPONENT_DIR):$(WORKSPACE_COMPONENT_DIR):delegated \
+        $(DOCKER_CREATE_OPTS) make resolve-local
+```
+  as you can see, rule `resolve-local` is executed inside container. 
+- if `resolve` pass, then the next rule will be executed.
+- if exit code of any rules is different than 0, makefile will abort execution and fail.
+
+List of available BUILDPACK_FUNCTIONS, which generates target dynamically:
+- buildpack-mount - create rule and mount component directory as volume
+- buildpack-cp-ro - create rule and copy components files to docker container.
+
+Target types:
+- `MOUNT_TARGET` - Contains rules which will be dynamically created by `buildpack-mount` function.
+- `COPY_TARGET` - Contains rules which will be dynamically created by `buildpack-cp-ro` function.
+
+## How to adjust makefile
+### Disable the current rule in the local Makefile
+To disable a rule in the new Makefile, follow it with the semicolon `;`.
+For example, write: `{RULE}: ;`.
+This results in the rule being disabled and warnings printed on the console.
+### How to add new local rule, which doesn't need `BUILDPACK`:
+Define rule in local makefile.
+Add this rule to one of the  global rule:
+```makefile
+verify:: own-rule
+```
+
+### How to add new rule in local makefile, which needs buildpack:
+Define rule in local makefile and call function which will create the rule:
+```makefile
+{my-rule}-local: 
+    do sth
+
+$(eval $(call {BUILDPACK_FUNCTION},my-rule)) # function which will create the new rule
```

Comment:
What is that? :)

Reply:
It's magic : )
This create the `{RULE}` which will execute `{RULE}-local` inside the docker.